miden-crypto/src/merkle/sparse_path.rs

```diff
@@ -0,0 +1,813 @@
+        }
+    }
+
+    /// Manually test the exact bit patterns for a sample path of 8 nodes, including both empty and
+    /// non-empty nodes.
+    ///
+    /// This also offers an overview of what each part of the bit-math involved means and
+    /// represents.
+    #[test]
+    fn test_sparse_bits() {
+        const DEPTH: u8 = 8;
+        let raw_nodes: [RpoDigest; DEPTH as usize] = [
+            // Depth 8.
+            ([8u8, 8, 8, 8].into()),
+            // Depth 7.
+            *EmptySubtreeRoots::entry(DEPTH, 7),
+            // Depth 6.
+            *EmptySubtreeRoots::entry(DEPTH, 6),
+            // Depth 5.
+            [5u8, 5, 5, 5].into(),
+            // Depth 4.
+            [4u8, 4, 4, 4].into(),
+            // Depth 3.
+            *EmptySubtreeRoots::entry(DEPTH, 3),
+            // Depth 2.
+            *EmptySubtreeRoots::entry(DEPTH, 2),
+            // Depth 1.
+            *EmptySubtreeRoots::entry(DEPTH, 1),
+            // Root is not included.
+        ];
+
+        let sparse_nodes: [Option<RpoDigest>; DEPTH as usize] = [
+            // Depth 8.
+            Some([8u8, 8, 8, 8].into()),
+            // Depth 7.
+            None,
+            // Depth 6.
+            None,
+            // Depth 5.
+            Some([5u8, 5, 5, 5].into()),
+            // Depth 4.
+            Some([4u8, 4, 4, 4].into()),
+            // Depth 3.
+            None,
+            // Depth 2.
+            None,
+            // Depth 1.
+            None,
+            // Root is not included.
+        ];
+
+        const EMPTY_BITS: u64 = 0b0110_0111;
+
+        let sparse_path = SparseMerklePath::from_sized_iter(raw_nodes).unwrap();
+
+        assert_eq!(sparse_path.empty_nodes_mask, EMPTY_BITS);
+
+        // Depth 8.
+        {
+            let depth: u8 = 8;
+
+            // Check that the way we calculate these indices is correct.
+            let idx = (sparse_path.depth() - depth) as usize;
+            assert_eq!(idx, 0);
+
+            // Check that the way we calculate these bitmasks is correct.
+            let bit = 0b1000_0000;
+            assert_eq!(bit, 1 << (depth - 1));
+
+            // Check that the depth-8 bit is not set...
+            let is_set = (sparse_path.empty_nodes_mask & bit) != 0;
+            assert!(!is_set);
+            // ...which should match the status of the `sparse_nodes` element being `None`.
+            assert_eq!(is_set, sparse_nodes.get(idx).unwrap().is_none());
+
+            // And finally, check that we can calculate non-empty indices correctly.
+            let control_node = raw_nodes.get(idx).unwrap();
+            let nonempty_idx: usize = 0;
+            assert_eq!(sparse_path.get_nonempty_index(NonZero::new(depth).unwrap()), nonempty_idx);
+            let test_node = sparse_path.nodes.get(nonempty_idx).unwrap();
+            assert_eq!(test_node, control_node);
+        }
+
+        // Rinse and repeat for each remaining depth.
```

Comment:
This is very explicit which is great on one hand, but on the other, is somewhat tedious to review, so if it's possible to deduplicate this, that'd be great. One idea is two for loops, one which checks all the depths that are empty and the other checks for non-empty nodes.